### ISSUE TITLE
How do we discriminate polymorphic types

### ISSUE THREAD

**Issue by ericprud** [w3c/hcls-fhir-rdf#138]
**Turtle filename**
* [Observation.valueCodeableConcept](https://build.fhir.org/observation-example-f206-staphylococcus.ttl.html)
* [Observation.valueSampleData](https://build.fhir.org/observation-example-sample-data.ttl.html)
* [Observation.valueString](https://build.fhir.org/observation-example-eye-color.ttl.html)
* [Observation.component.valueCodeableConcept](https://build.fhir.org/observation-example-alcohol-type.ttl.html)
* [Observation.valueString](https://build.fhir.org/observation-example-abdo-tender.ttl.html)

**FHIR Resource**
URL of corresponding attribute on build.fhir.org, e.g. https://build.fhir.org/observation.html#a26.b

[Observation.instantiates](https://build.fhir.org/observation#a1.)
[Observation.subject](https://build.fhir.org/observation#a14.)

**ShEx shape**
[`@<Observation>~fhir:instantiates`](https://build.fhir.org/observation.shex.html)

**Description**
Current definition of polymorphic types requires addition of either
* complex type: `rdf:type` arc
* primitive type: `^^xsd:datatype` scalar type annotation

Is e.g. `xsd:date` a representation of more than one primitive type? Are there any polymorphic properties that include those primitive types? (Will there ever be?)

What about primitive types that are `xsd:string`s?

 à la:
``` diff
  <Observation/123>
    #  instantiates Reference
    fhir:instantiates [
+      a fhir:Reference ; # type needed because fhir:instantiates is polymorphic
       fhir:link <http://example.org/fhir/PlanDefinition/KDN5> ; # current
       fhir:reference [
         fhir:v "http://example.org/fhir/PlanDefinition/KDN5"^^xsd:anyURI # note: no version on reference
      ]
    ] ; 
    #  instantiates canonical
    fhir:instantiates [
+      a fhir:canonical ; # type needed because fhir:instantiates is polymorphic
       fhir:link <http://example.org/fhir/PlanDefinition/KDN5?version=v123> ; # new with this proposal
       fhir:v "http://example.org/fhir/PlanDefinition/KDN5|v123"^^xsd:anyURI ;
    ] ; # 
```

**Comment by dbooth-boston** [comment]
AGREED to add type arc on today's call.  See: https://www.w3.org/2024/01/04-hcls-irc#T16-46-55

ACTION: DBooth to draft edits of the rdf.html page

**Comment by dbooth-boston** [comment]
Reopened pending the merge of the actual fix.

**Comment by dbooth-boston** [comment]
Needs to be fixed in HAPI.  @balhoff  will work on it.

**Comment by dbooth-boston** [comment]
The xsd:anyURI definition [seems to allow IRIs](https://www.w3.org/TR/xmlschema11-2/#anyURI), but AFAICT http://example.org/fhir/PlanDefinition/KDN5|v123 is still not technically permissible because the pipe character is not syntactically allowed by RFC 3986/3987.  However, xs:anyURI is the datatype shown by the FHIR spec, so this inconsistency seems to be a general FHIR issue rather than a specifically FHIR RDF issue.

**Comment by dbooth-boston** [comment]
@balhoff does https://github.com/hapifhir/org.hl7.fhir.core/pull/1853 also fix this issue?

**Comment by balhoff** [comment]
@dbooth-boston it seems like it would.

**Comment by dbooth-boston** [comment]
@balhoff , does this require any change to the ontology, or does the ontology already indicate all of these types?

@ericprud , it looks to me (in [this example](https://build.fhir.org/observation.shex.html)) like no ShEx changes need to be made for this decision, because the ShEx already indicates the FHIR type rather than the XSD type.  Is that correct?

**Comment by tmprd** [comment]
This is already shown in yellow templates
<img width="799" height="215" alt="Image" src="https://github.com/user-attachments/assets/cd4e764c-44f4-4547-bb4e-9cb1f8715d99" />
Also shown even for "open types", like Task.output.value, which can be anything https://hl7.org/fhir/task.html
<img width="395" height="721" alt="Image" src="https://github.com/user-attachments/assets/3e98b305-cc58-4d01-b642-5f9985db82a8" />

Also already shown in OWL:
"Open type" shows Element
<img width="344" height="223" alt="Image" src="https://github.com/user-attachments/assets/3f591624-84c1-4255-ad55-0ad42f6197a1" />

Choice types shown (but broken due to #178 and incomplete due to #177 
<img width="612" height="57" alt="Image" src="https://github.com/user-attachments/assets/74781639-60dd-4f57-8fd2-f74781ee6bc4" />

**Comment by dbooth-boston** [comment]
~~@detlef-grittner-sohard~~ @ericprud  , what is the status of fixing this in HAPI for R6?    Note that it was already fixed in the FHIR R5 "core library", which I believe is used both by HAPI and by the FHIR spec build process: https://github.com/hapifhir/org.hl7.fhir.core/pull/1853 , already merged.

(EDIT: Detlef replied that EricP had taken over HAPI.)

**Comment by dbooth-boston** [comment]
Closing, as fixed by https://github.com/HL7/kindling/pull/200
